Reconfigure step (#2099)
home-assistant/developers.home-assistant: docs/config_entries_config_flow_handler.md

    return True
```

## Reconfigure

> A config entry can allow reconfiguration by adding a `reconfigure` step. This provides a way for integrations to allow users to change config entry data without the need to implement an `OptionsFlow` for changing setup data which are not meant to be optional.

Reviewer: ```suggestion
A config entry can allow reconfiguration by adding a `reconfigure` step. This provides a way for integrations to allow users to change config entry data without the need to implement an `OptionsFlow` for changing setup data which is not meant to be optional.
```

> This is not meant to handle authentication issues or reconfiguration of such. For that we have the [`reauth`](#reauthentication) step which should be implemented.

Reviewer: There's no way to start a reauth step by the user.

Author: That's not the intention it's implied it can be manually started. I'll extend the sentence a bit to clarify.

> A config entry can allow reconfiguration by adding a `reconfigure` step. This provides a way for integrations to allow users to change config entry data without the need to implement an `OptionsFlow` for changing setup data which is not meant to be optional.
> 
> This is not meant to handle authentication issues or reconfiguration of such. For that we have the [`reauth`](#reauthentication) step which should be implemented to automatically start in such case there is an issue with authentication.

Reviewer 2: ```suggestion
A config entry can allow reconfiguration by adding a `reconfigure` step. This provides a way for integrations to allow users to change config entry data without the need to implement an `OptionsFlow` for changing setup data which is not meant to be optional.

This is not meant to handle authentication issues or reconfiguration of such. For that we have the [`reauth`](#reauthentication) step, which should be implemented to automatically start in such case there is an issue with authentication.
```

```python
import voluptuous as vol

class ExampleConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
    """Config flow for Example integration."""

>     async def async_step_reconfigure(self, user_input):

Reviewer: Please add type annotations.

        if user_input is not None:
            pass  # TODO: process user input

        return self.async_show_form(
            step_id="reconfigure",
            data_schema=vol.Schema({vol.Required("input_parameter"): str}),
        )
```

## Reauthentication

Gracefully handling authentication errors such as invalid, expired, or revoked tokens is needed to advance on the [Integration Qualily Scale](integration_quality_scale_index.md). This example of how to add reauth to the OAuth flow created by `script.scaffold` following the pattern in [Building a Python library](api_lib_auth.md#oauth2).

This example catches an authentication exception in config entry setup in `__init__.py` and instructs the user to visit the integrations page in order to reconfigure the integration.

> To change config entry data which is not optional (`OptionsFlow`) but is also not to handle authentication we have the [`reconfigure`](#reauthentication) step which can be implemented and should not be handled in a `reauth` step.

Reviewer 3: ```suggestion
To allow the user to change config entry data which is not optional (`OptionsFlow`) and not directly related to authentication, for example a changed host name, integrations should implement the [`reconfigure`](#reauthentication) step.
```

    return True
```

> ## Reconfigure

Reviewer 3: We should make sure the difference between reconfigure and reauthenticate is clear.

Author: That whole page needs a bit of an overhaul but let's do that in a follow-up 👍
